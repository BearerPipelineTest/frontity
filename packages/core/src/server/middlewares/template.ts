import { Middleware, Next } from "koa";
import { Context } from "@frontity/types";
import { Template } from "../../../types";

/**
 * The template function to return the html.
 *
 * @param options - The options to compile the template with.
 *
 * @returns The compiled html with the options.
 */
const html: Template = ({
  html,
  scripts,
  head,
  htmlAttributes,
  bodyAttributes,
}) => `<!doctype html>
    <html ${htmlAttributes || ""}>
      <head>
        <meta charset="utf-8">
        <meta name="generator" content="Frontity">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        ${head ? head.join("\n") : ""}
      </head>
      <body ${bodyAttributes || ""}>
        <div id="root">${html}</div>
        ${scripts ? scripts.join("\n") : ""}
      </body>
    </html>`;

/**
 * Defines the module stats for the current request.
 *
 * @param ctx - Koa context.
 * @param next - The next method.
 *
 * @returns The awaited next mehtod.
 */
export const template = async (
  ctx: Context,
  next: Next
): Promise<Middleware> => {
  const { store } = ctx.state;

  ctx.state.template = html;

  // Get the correct template or html if none is found.
  if (store.libraries.frontity.template) {
    // Get a reference for the previous, default html template.
    const defaultTemplate = ctx.state.template;

    // Define the new template method.
    ctx.state.template = (args: any) => {
      return store.libraries.frontity.template({ ...args, defaultTemplate });
    };
  }

  return await next();
};
