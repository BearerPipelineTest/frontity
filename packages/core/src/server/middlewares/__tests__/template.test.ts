import { template } from "../template";
import { createKoaContext } from "./__utilities__/create-koa-context";
import { fakeNext } from "./__utilities__/fake-next";

describe("template", () => {
  it("should define a default template", async () => {
    const ctx: any = createKoaContext();

    await template(ctx, fakeNext);

    expect(ctx.state.template.toString()).toMatchSnapshot();
  });

  it("the default template should accept head, scripts and attributes", async () => {
    const ctx: any = createKoaContext();

    await template(ctx, fakeNext);

    expect(
      ctx.state.template({
        head: [`<custom-tag attribute="value" />`],
        scripts: [`<script src="/path/to/script"></script>`],
        htmlAttributes: 'lang="en"',
        bodyAttributes: 'class="custom"',
      })
    ).toMatchSnapshot();
  });

  describe("custom template", () => {
    it("should define a custom template", async () => {
      const ctx: any = createKoaContext({
        template: jest.fn(),
      });

      // Call the middlware.
      await template(ctx, fakeNext);

      // The new template method should be redefined.
      expect(ctx.state.template.toString()).toMatchSnapshot();
    });

    it("should be able to call defaultTemplate", async () => {
      const ctx: any = createKoaContext({
        template: ({ defaultTemplate }) => {
          return defaultTemplate({
            head: ['<meta rel="custom" />'],
          });
        },
      });

      // Call the middleware.
      await template(ctx, fakeNext);

      // Expect the template to have a custom meta tag.
      expect(ctx.state.template()).toMatchSnapshot();
    });
  });
});
