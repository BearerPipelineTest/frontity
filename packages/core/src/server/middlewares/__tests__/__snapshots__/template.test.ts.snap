// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template custom template should be able to call defaultTemplate 1`] = `
"<!doctype html>
    <html >
      <head>
        <meta charset=\\"utf-8\\">
        <meta name=\\"generator\\" content=\\"Frontity\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        <meta rel=\\"custom\\" />
      </head>
      <body >
        <div id=\\"root\\">undefined</div>
        
      </body>
    </html>"
`;

exports[`template custom template should define a custom template 1`] = `
"(args) => {
            return store.libraries.frontity.template(Object.assign(Object.assign({}, args), { defaultTemplate }));
        }"
`;

exports[`template should define a default template 1`] = `
"({ html, scripts, head, htmlAttributes, bodyAttributes, }) => \`<!doctype html>
    <html \${htmlAttributes || \\"\\"}>
      <head>
        <meta charset=\\"utf-8\\">
        <meta name=\\"generator\\" content=\\"Frontity\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        \${head ? head.join(\\"\\\\n\\") : \\"\\"}
      </head>
      <body \${bodyAttributes || \\"\\"}>
        <div id=\\"root\\">\${html}</div>
        \${scripts ? scripts.join(\\"\\\\n\\") : \\"\\"}
      </body>
    </html>\`"
`;

exports[`template the default template should accept head, scripts and attributes 1`] = `
"<!doctype html>
    <html lang=\\"en\\">
      <head>
        <meta charset=\\"utf-8\\">
        <meta name=\\"generator\\" content=\\"Frontity\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        <custom-tag attribute=\\"value\\" />
      </head>
      <body class=\\"custom\\">
        <div id=\\"root\\">undefined</div>
        <script src=\\"/path/to/script\\"></script>
      </body>
    </html>"
`;
